src/system-plugins/input-controller/public/js/input-controller.js

```diff
@@ -1,239 +1,513 @@
-    self.registerdCommands = {};
-    self.registerdControls = {};
-    self.controllers = [];
-    // add our known controllers
-    self.controllers.push(new inputController.Keyboard(cockpit));
-    self.controllers.push(new inputController.Gamepad(cockpit));
-    self.checkDuplicates = function () {
-      var commandBindings = [];
-      var duplicateInformation = [];
-      for (var command in self.registerdCommands) {
-        if (self.registerdCommands[command].active) {
-          for (var binding in self.registerdCommands[command].bindings) {
-            commandBindings.push({
-              value: binding + ':' + self.registerdCommands[command].bindings[binding],
-              name: self.registerdCommands[command].name
-            });
-          }
+  inputController.InputController = class InputController
+  {
+    constructor(cockpit)
+    {
+      console.log("Starting Input Controller");
+      this.cockpit = cockpit;
+      this.currentPreset = "OpenROVDefault";
+      
+      this.deferSetupForMousetrap();
+    };
+
+    deferSetupForMousetrap()
+    {
+        var self = this;
+
+        if ( (typeof Mousetrap !== "undefined") && recordPluginNeeded) {
+          loadScript('components/mousetrap-js/plugins/record/mousetrap-record.js');
+          recordPluginNeeded=false;
         }
-      }
-      commandBindings.forEach(function (binding) {
-        commandBindings.forEach(function (checkBinding) {
-          if (binding === checkBinding)
-            return;
-          if (binding.value === checkBinding.value) {
-            duplicateInformation.push('Command name: \'' + binding.name + '\' Binding: ' + binding.value);
+        
+        if((typeof Mousetrap == "undefined") || (typeof Mousetrap.record == "undefined") || self.setup() == false )
+        {          
+          // Call timeout again
+          setTimeout( self.deferSetupForMousetrap.bind( self ), 100 );
+        }
+    }
+
+    setup()
+    {
+      var self = this;
+
+      if (typeof Mousetrap !== "undefined")
+      {
+
+        self.mousetrap = Mousetrap;
+        self.controllers = new Map();
+        //Data structure to hold our presets on the browser side
+        //key: "preset-name" value: ["preset1", "preset2", ..., "presetN"]
+        self.presets = new Map();
+
+        //The OpenROV default mapping preset. Defined by crawling the plugins
+        self.openrovPreset = new inputController.Preset("OpenROV Preset");
+
+        //Add our default controllers
+        self.keyboard = new Keyboard(cockpit, self.mousetrap);
+        self.controllers.set("keyboard", self.keyboard);
+
+        self.gamepad = new Gamepad(cockpit);
+        self.controllers.set("gamepad", self.gamepad);
+
+        //Add the default preset to the map
+        self.presets.set("OpenROVDefault", self.openrovPreset);
+
+        //If plugins have loaded before us, let's get their defaults
+        self.cockpit.loadedPlugins.forEach(function(plugin) {
```

Comment:
From an activation stand point. As long as this code executes in the listen() method, it is guaranteed that all other plugins have finished their constructors and are in the loadedPlugins array.  So no concerns about something loading before the input-controller plugin.